### PR TITLE
authz: add Perforce provider

### PR DESCRIPTION
**Pull request by unknwon**
This PR adds the `authz.Provider` implementation for Perforce.

The code will not yet be in effect when merge, upcoming PR will initialize Perforce authz provider.

**NOTE:** The `go.sum` change come out of nowhere 🤷‍♂️ 

Subset of #17881 and extracted here for easier review, and part of #16705.

### CODE REVIEW

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go [hunk 1]
```diff
@@ -197,6 +197,21 @@ func (s *PermsSyncer) syncUserPerms(ctx context.Context, userID int32, noPerms b
...
+		database.UserEmailsListOptions{
+			UserID:       user.ID,
+			OnlyVerified: true,
+		},
+	)
+	if err != nil {
+		return errors.Wrap(err, "list user verified emails")
+	}
+
+	emails := make([]string, len(userEmails))
```
Comment (ryanslade) [inline enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:210]: Is it an issue if the user has zero verified emails?

File: internal/authz/perforce/perforce.go [hunk 2]
```diff
@@ -0,0 +1,483 @@
...
+			}
+
+		} else {
+			if !p.canGrantReadAccess(level) {
+				continue
+			}
+
+			includePrefixes = append(includePrefixes, extsvc.RepoID(depotPrefix))
+		}
+	}
```
Comment (ryanslade) [inline internal/authz/perforce/perforce.go:243]: I think you forget to check `scanner.Err()` here

**Review by ryanslade — APPROVED** [review]

File: enterprise/cmd/repo-updater/internal/authz/perms_syncer.go (hunk 1 above)
Comment (unknwon) [reply enterprise/cmd/repo-updater/internal/authz/perms_syncer.go:210]: In that case, the individual authz.Provider implementation may decide return nothing (ie. Perforce)

File: internal/authz/perforce/perforce.go (hunk 2 above)
Comment (unknwon) [reply internal/authz/perforce/perforce.go:243]: Nice catch!

Comment (unknwon) [reply internal/authz/perforce/perforce.go:243]: Oh no... I just realized I checked it on the return statement. But apparently that's not obvious, I'll separate it out.

**Pull request merged**